### PR TITLE
Rule sampler

### PR DESCRIPTION
**Pull request by marcotc**
This PR adds support for rule-based sampling of traces.

In its most complete form, here's an example configuration:

```ruby
Datadog.configure do |c|
  c.tracer sampler: Datadog::PrioritySampler.new(
    post_sampler: Datadog::Sampling::RuleSampler.new(
      [
        Datadog::Sampling::SimpleRule.new(name: 'operation.name', sample_rate: 0.9),
        Datadog::Sampling::SimpleRule.new(service: 'service-1', sample_rate: 0.5),
        Datadog::Sampling::SimpleRule.new(service: /service-.*/, name: 'db.select', sample_rate: 0.7),
      ],
      default_sampler: Datadog::RateSampler.new(1.0),
      rate_limiter: Datadog::Sampling::TokenBucket.new(1000),
    )
  )
end
```

Or for 100% sampling with 100 traces/sec limit:
```ruby
Datadog.configure do |c|
  c.tracer sampler: Datadog::PrioritySampler.new(
    post_sampler: Datadog::Sampling::RuleSampler.new
  )
end
```

### CODE REVIEW

File: lib/ddtrace/sampling/matcher.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      # @return [Boolean]
+      def match?(span)
+        raise NotImplementedError
+      end
+    end
+
+    class SimpleMatcher < Matcher
+      # Returns `true` for case equality (===) with any object
+      MATCH_ALL = Class.new do
+        # DEV: A class that implements `#===` is ~20% faster than
```
Comment (brettlangdon) [inline lib/ddtrace/sampling/matcher.rb]: have a link? this is neat.

File: lib/ddtrace/sampling/rule.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+      # If `sampler` accepts the `span`, this method returns an {Array}
+      # with the sampling decision {Boolean} and a sampling rate {Float}.
+      #
+      # If the concept of sampling rate does not apply to the `sampler`
+      # `nil` is returned as the sample rate.
+      #
+      # @param [Span] span
+      # @return [Array<Boolean, Float>] sampling decision and sampling rate
+      # @return [NilClass] if this rule does not apply
+      #   or `nil` if this rule does not apply
```
Comment (brettlangdon) [inline lib/ddtrace/sampling/rule.rb]: Not sure this `@return` description makes sense

File: lib/ddtrace/sampling/rule_sampler.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+module Datadog
+  module Sampling
+    # TODO:
+    class RuleSampler
+      extend Forwardable
+
+      attr_reader :rules, :rate_limiter, :priority_sampler
+
+      def initialize(rules, rate_limiter, priority_sampler = Datadog::RateByServiceSampler.new)
```
Comment (brettlangdon) [inline lib/ddtrace/sampling/rule_sampler.rb]: we would need to make sure this priority sampler gets it's rates updated from responses from the agent.

File: lib/ddtrace/sampling/rule_sampler.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      def sample?(span)
+        sampled, _ = sample_span(span) { |s| return @priority_sampler.sample?(s) }
+        sampled
+      end
+
+      def sample!(span)
+        sampled, sample_rate = sample_span(span) { |s| return @priority_sampler.sample!(s) }
+
+        # Set metrics regardless of sampling outcome
+        set_metrics(span, sample_rate, rate_limiter.effective_rate)
```
Comment (brettlangdon) [inline lib/ddtrace/sampling/rule_sampler.rb]: we only want to set the rate if the sampler had a hand in making the decision.

e.g. if the rate limiter doesn't get evaluated, then we don't set the metric for the rate limiter.

Can think of this as sort of debugging for why the span was sampled or not. e.g. if the rule decides not to sample, then rate limiter didn't play a role in making that decision, we can see that by the rate limit psr metric not being set.

File: lib/ddtrace/sampling/rule_sampler.rb (hunk 3 above)
Comment (marcotc) [reply lib/ddtrace/sampling/rule_sampler.rb]: That's taken take by delegating `:update` to that sampler, like so further down this class:
```ruby
def_delegators :@priority_sampler, :update
```

File: lib/ddtrace/sampling/matcher.rb (hunk 1 above)
Comment (marcotc) [reply lib/ddtrace/sampling/matcher.rb]: I tested it locally while developing this class, I can't upload a gist a benchmark snippet.

File: lib/ddtrace/sampling/rule.rb (hunk 2 above)
Comment (marcotc) [reply lib/ddtrace/sampling/rule.rb]: Addressed, remove out of place comment line.

File: lib/ddtrace/sampling/rule_sampler.rb (hunk 4 above)
Comment (marcotc) [reply lib/ddtrace/sampling/rule_sampler.rb]: Got it, I'll address to match this requirement.

Comment (marcotc) [reply lib/ddtrace/sampling/rule_sampler.rb]: I've made changes to only set rule based metrics when rules apply to a trace.
The same goes to the rate limiter: only set metrics when the limiter was evaluated.

File: draft.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+      Datadog::Sampling::SimpleRule.new(name: 'operation.name', sample_rate: 0.9),
+      Datadog::Sampling::SimpleRule.new(service: 'service-1', sample_rate: 0.9),
+      Datadog::Sampling::SimpleRule.new(sample_rate: 0.7) { |span| span.name != 'important.operation' },
+      Datadog::Sampling::SimpleRule.new(sample_rate: 1.0),
+    ],
+    Datadog::Sampling::TokenBucket.new(1),
+    Datadog::PrioritySampler.new(
+      post_sampler: Datadog::RateByServiceSampler.new(
+        1.0,
+        env: proc { Datadog.tracer.tags[:env] } # TODO how do I provide `tracer.tags`? Seems like a circular reference here.
```
Comment (marcotc) [inline draft.rb]: Initializing a fallback `PrioritySampler` is not trivial.
Maybe we could use an interface in our tracer that allows us to append or prepend a sampler to a sampling chain, so that we could prepend our `RuleSampler` before the existing `PrioritySampler`, and avoid having all this complex initialization information outside of the tracer.

File: draft.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+      Datadog::Sampling::SimpleRule.new(name: 'operation.name', sample_rate: 0.9),
+      Datadog::Sampling::SimpleRule.new(service: 'service-1', sample_rate: 0.9),
+      Datadog::Sampling::SimpleRule.new(sample_rate: 0.7) { |span| span.name != 'important.operation' },
+      Datadog::Sampling::SimpleRule.new(sample_rate: 1.0)
+    ],
+    Datadog::Sampling::TokenBucket.new(1),
+    Datadog::PrioritySampler.new(
+      post_sampler: Datadog::RateByServiceSampler.new(
+        1.0,
+        env: proc { Datadog.tracer.tags[:env] } # TODO: how do I provide `tracer.tags`? Seems like a circular reference here.
```
Comment (marcotc) [inline draft.rb]: Initializing a fallback `PrioritySampler` is not trivial.
Maybe we could use an interface in our tracer that allows us to append or prepend a sampler to a sampling chain, so that we could prepend our `RuleSampler` before the existing `PrioritySampler`, and avoid having all this complex initialization information outside of the tracer.

File: lib/ddtrace/sampling/token_bucket.rb [hunk 7]
```diff
@@ -98,5 +98,21 @@ def increment_conforming_count
         @conforming_messages += 1
       end
     end
+
+    # TODO: This class name is so bad, yet so good.
+    # [Class documentation]
+    class UnlimitedLimiter < RateLimiter
```
Comment (delner) [inline lib/ddtrace/sampling/token_bucket.rb]: Haha, great name. If you want a different name my suggestions are just as quirky: `NonLimiter`, `NoLimiter`, `NeverLimiter`. But honestly, `UnlimitedLimiter` works just fine.

I might suggest though that this behavior for a limiter shouldn't be a `Limiter` but a composable module for a `Limiter` (e.g. `RateLimiiter.new.extend(UnlimitedRate)`) because an unlimited limiter isn't a distinct species of limiter, just one that behaves a particular way.

File: lib/ddtrace/sampling/matcher.rb (hunk 1 above)
Comment (delner) [reply lib/ddtrace/sampling/matcher.rb]: Interesting idea; I could benchmark it and post results.

File: lib/ddtrace/sampling/rule_sampler.rb [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        span.sampled = sampled
+        sampled
+      end
+
+      def_delegators :@priority_sampler, :update
+
+      private
+
+      def sample_span(span)
+        sampled, sample_rate = @rules.find do |rule|
```
Comment (delner) [inline lib/ddtrace/sampling/rule_sampler.rb]: This loop is a little awkward because you have to find and retrieve two different values from the rule. It may be possible to eliminate `Rule#sample` entirely, and simplify things. If we were to change `Rule` to delegate to `Matcher#match?`, `Sampler#sample?`, `Sampler#sample_rate`, then you could rewrite this function as:

```ruby
def sample_span(span)
  rule = @rules.find do |rule|
    rule.matches?(span)
  end

  yield(span) unless rule

  sampled = rule.sample?(span)
  sample_rate = rule.sample_rate(span)
  [sampled && rate_limiter.allow?(1), sample_rate]
end
```

As a general rule of thumb, it might be better for components like `Rule` to have very simple methods that do as little as possible and have more complex, compositional objects such as this `RuleSampler` orchestrate and drive those smaller components.

There might be even better ways of doing this than what I've suggested, but just some food for thought.

File: lib/ddtrace/sampler.rb [hunk 9]
```diff
@@ -96,6 +107,8 @@ def sample_rate(span)
 
       @mutex.synchronize do
         @sampler.fetch(key, @fallback).sample_rate
+        # TODO: set sampling metric here, for fallback case
```
Comment (brettlangdon) [inline lib/ddtrace/sampler.rb]: We want to set this metric anytime this sampler is used, not just for "fallback"

File: lib/ddtrace/sampler.rb [hunk 10]
```diff
@@ -141,10 +154,19 @@ def sample!(span)
...
           #       be sent to the agent. Otherwise metrics for traces will not be accurate, since the
           #       agent will have an incomplete dataset.
-          priority = priority_sample(span) ? Datadog::Ext::Priority::AUTO_KEEP : Datadog::Ext::Priority::AUTO_REJECT
+          #
+          #       We also ensure that the sampling rate is not propagated to the agent, to avoid erroneous
+          #       metric upscaling.
+          span.sampled = true
+          # TODO: do i have to unset SAMPLE_RATE_METRIC_KEY? or setting it to 1.0 is fine?
+          # TODO: a method to unset doesn't exist yet.
+          span.set_metric(SAMPLE_RATE_METRIC_KEY, 1.0) if span.get_metric(SAMPLE_RATE_METRIC_KEY)
```
Comment (brettlangdon) [inline lib/ddtrace/sampler.rb]: This should only be set on "RateSampler" a.k.a "client sampling".

For a sampler which sets `span.sampled = false` we need to set this metric to tell the agent how to upscale metrics for the traces that are not sent to it.

For "priority sampling" every span is sampled and sent to the agent, so we should not set this.

File: lib/ddtrace/sampling/rule_sampler.rb [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    # TODO: Write class documentation
+    # RuleSampler
+    class RuleSampler
+      extend Forwardable
+
+      attr_reader :rules, :rate_limiter, :fallback_sampler
+
+      def initialize(rules = [],
+                     rate_limiter = Datadog::Sampling::UnlimitedLimiter.new,
+                     fallback_sampler = Datadog::AllSampler.new)
```
Comment (brettlangdon) [inline lib/ddtrace/sampling/rule_sampler.rb]: This isn't really a "fallback" as it is a "default"

So, we should probably just define this as `default_sample_rate = 1.0`

Then we do `@default_sampler = RateSampler. new(default_sample_rate)` instead of `@fallback_sampler = ...`

File: lib/ddtrace/sampling/rule_sampler.rb [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+        return yield(span) if rule.nil?
+
+        sampled = rule.sample?(span)
+        sample_rate = rule.sample_rate(span)
+
+        set_rule_metrics(span, sample_rate)
+
+        return false unless sampled
+
+        rate_limiter.allow?(1).tap do
```
Comment (brettlangdon) [inline lib/ddtrace/sampling/rule_sampler.rb]: remind me Ruby syntax, the `.tap` block will always get executed here, regardless of the return value of `rate_limiter.allow?(1)`?

File: lib/ddtrace/sampling/rule_sampler.rb [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+      # Use {#sample!} instead
+      def sample?(_span)
+        raise 'RuleSampler cannot be evaluated without side-effects'
+      end
+
+      def sample!(span)
+        sampled = sample_span(span) { |s| @fallback_sampler.sample!(s) }
+
+        sampled.tap do
+          span.sampled = sampled
```
Comment (brettlangdon) [inline lib/ddtrace/sampling/rule_sampler.rb]: for here, we always want `span.sampled = true`, we want to do:

```ruby
span.sampled = true
span.context.sampling_priority = sampled ? Datadog::Ext::Priority::AUTO_KEEP : Datadog::Ext::Priority::AUTO_REJECT
```

File: lib/ddtrace/sampling/rule_sampler.rb [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+        end
+      end
+
+      def_delegators :@fallback_sampler, :update
```
Comment (brettlangdon) [inline lib/ddtrace/sampling/rule_sampler.rb]: we can `no-op` `:update`, we don't need to consider agent returned sample rates at all here.

File: lib/ddtrace/sampling/token_bucket.rb [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+      # Checks if a message of provided +size+
+      # conforms with the current bucket limit.
+      #
+      # If it does, return +true+ and remove +size+
+      # tokens from the bucket.
+      # If it does not, return +false+ without affecting
+      # the tokens form the bucket.
+      #
+      # @return [Boolean] +true+ if message conforms with current bucket limit
+      def allow?(size)
```
Comment (brettlangdon) [inline lib/ddtrace/sampling/token_bucket.rb]: you could short circuit for `return false if @rate == 0`

File: lib/ddtrace/sampling/rule_sampler.rb (hunk 12 above)
Comment (marcotc) [reply lib/ddtrace/sampling/rule_sampler.rb]: `value.tap{}` returns `value` regardless of what the block provided to `tap` does.
We normally use it to "inspect" the value, but still return the original value.

File: lib/ddtrace/sampling/rule_sampler.rb (hunk 11 above)
Comment (marcotc) [reply lib/ddtrace/sampling/rule_sampler.rb]: I think receiving a sampler implementation as the default fallback works better with how we structured our samplers.

The same functionality we can get from having a `default_sample_rate` can be achieved by receiving the sampler instance instead.

I do like the term `default` more than `fallback` though, so I'm renaming the parameter with the suggestion.

File: lib/ddtrace/sampling/rule_sampler.rb (hunk 14 above)
Comment (marcotc) [reply lib/ddtrace/sampling/rule_sampler.rb]: This is effectively a no-op, as most sampler implementations we delegate this call to are no-ops.

I think we should keep this delegate as we currently support `RateByServiceSampler` in our app. We could remove `update` it if we drop support for it in the future.

File: lib/ddtrace/sampling/token_bucket.rb (hunk 15 above)
Comment (marcotc) [reply lib/ddtrace/sampling/token_bucket.rb]: Done.

File: lib/ddtrace/sampling/rule_sampler.rb (hunk 11 above)
Comment (brettlangdon) [reply lib/ddtrace/sampling/rule_sampler.rb]: I think it might actually cause some configuration fatigue if we allow them to pass in any sampler.

Mostly because 1) they shouldn't and 2) we want them to be able to set the default sample rate if no rules match (very common case), if they have to change from an `AllSampler.new` to a `RateSampler.new(rate)`, then it is more that they need to know about vs `DD_DEFAULT_SAMPLE_RATE=0.5` for example.

File: lib/ddtrace/sampling/rule_sampler.rb (hunk 12 above)
Comment (brettlangdon) [reply lib/ddtrace/sampling/rule_sampler.rb]: perfect

File: lib/ddtrace/sampling/rule_sampler.rb (hunk 11 above)
Comment (marcotc) [reply lib/ddtrace/sampling/rule_sampler.rb]: The part where internal vs external APIs diverge in my opinion.

I think the internal API for the `RuleSampler` matches our inner-workings at the correct level.
But I do agree that for external facing configuration it's just too cumbersome.

File: lib/ddtrace/sampler.rb (hunk 10 above)
Comment (marcotc) [reply lib/ddtrace/sampler.rb]: Would this mean that sending `1.0` on `_sample_rate` for a sampler that always sets `span.sampled = true` is correct?

File: lib/ddtrace/sampling/rule_sampler.rb (hunk 11 above)
Comment (marcotc) [reply lib/ddtrace/sampling/rule_sampler.rb]: Addressed this. We now have a more humane API :)
```ruby
RuleSampler.new(rules, # Array of rules here
  rate_limit: 1000,
  default_sample_rate: 1.0)
```

File: lib/ddtrace/sampler.rb [hunk 16]
```diff
@@ -175,8 +199,8 @@ def priority_assigned_upstream?(span)
       span.context && !span.context.sampling_priority.nil?
     end
 
-    def priority_sample(span)
-      @priority_sampler.sample?(span)
+    def priority_sample!(span)
+      @priority_sampler.sample!(span)
```
Comment (marcotc) [inline lib/ddtrace/sampler.rb]: We change this to use `sample!` instead of `sample?` to allow for `@priority_sampler` to set additional tags that are unrelated to priority sampling.
In our case, we want to set '_dd.rule_psr' and  '_dd.limit_psr' when sampling happens.

File: lib/ddtrace/sampling/rule_sampler.rb (hunk 8 above)
Comment (marcotc) [reply lib/ddtrace/sampling/rule_sampler.rb]: There is one trade-off with this approach which is that custom rules might want to provide a coupled result between the sampling decision (`sample?`) and sample_rate. In this case we are making it harder to implement such rules:
```ruby
class CustomRule < Rule
  def sample?(span)
    case span.service
    when 'service-1'
      span.name != 'ignore.op'
    when 'service-2'
      span.start_time.hour > 7
    end
  end

  def sample_rate(span)
    case span.service
    when 'service-1'
      0.8
    when 'service-2'
      0.9
    end
  end
end
```

Instead of returning a "response" payload with both together:
```ruby
Rule.new do |span|
  case span.service
  when 'service-1'
    [span.name != 'ignore.op', 0.8]
  when 'service-2'
    [span.start_time.hour > 7, 0.9]
  end
end
```
After reviewing this, it seems like an ergonomic decision that we wouldn't prioritize in detriment of the maintainability of `RuleSampler`. Thus I made changes that mostly follow your suggestion.

File: lib/ddtrace/sampler.rb (hunk 9 above)
Comment (marcotc) [reply lib/ddtrace/sampler.rb]: 👌, this was a reminder so that we don't forget to add a new metric for the existing `RateByServiceSampler`. This going to be addressed in a separate PR.

File: lib/ddtrace/sampler.rb (hunk 10 above)
Comment (marcotc) [reply lib/ddtrace/sampler.rb]: This has been properly addressed: the tag is only set we are pre-sampling in memory and not sending all spans to the agent, as it has always been.

File: lib/ddtrace/sampling/rule_sampler.rb (hunk 13 above)
Comment (marcotc) [reply lib/ddtrace/sampling/rule_sampler.rb]: Our samplers can be used as stand-alone samplers, as well as composed together.
When using `RuleSampler` as a stand-alone sampler, it will make hard decisions.

When using it alongside the `PrioritySampler`, the sampling_priority and final sampling decision are taken care of by the priority sampler, which will enforce the conditions you mentioned.

`PrioritySampler` is currently the sampler in our code base that is aware of `sampling_priority`. All other samplers only make a boolean decision and are not concerned with details around what flags to set in the `span`.

File: lib/ddtrace/sampling/token_bucket.rb (hunk 7 above)
Comment (marcotc) [reply lib/ddtrace/sampling/token_bucket.rb]: I do see the point you are making, specially if we had something like a `DebugRateLimiter` that prints all operations to stdout: we would be able to add this functionality to an existing rate limiter as a mixin.

But for the case of `UnlimitedLimiter`, I think that being subclass of `RateLimiter` makes more sense, as it pretty much overrides all behaviour of `RateLimiter` and replaces with its own. I see it less as a composable feature-set and more as a whole type of limiter in itself. When overlaying it with the token bucket implementation, for example, it removes all token bucket functionality and replaces with its own implementation.

File: lib/ddtrace/sampler.rb [hunk 17]
```diff
@@ -194,11 +204,24 @@ def sample!(span)
...
-          priority = priority_sample(span) ? Datadog::Ext::Priority::AUTO_KEEP : Datadog::Ext::Priority::AUTO_REJECT
+          #
+          #       We also ensure that the agent knows we that our `post_sampler` is not performing true sampling,
+          #       to avoid erroneous metric upscaling.
+          span.sampled = true
+          if pre_sample_rate_metric
+            span.set_metric(SAMPLE_RATE_METRIC_KEY, pre_sample_rate_metric) # Restore true sampling metric
+          else
+            span.clear_metric(SAMPLE_RATE_METRIC_KEY)
+          end
```
Comment (marcotc) [inline lib/ddtrace/sampler.rb]: This class is growing in complexity in order to support the `RuleSampler` as a `post_sampler`.
It seems nice to keep all priority sampling concerns here, and not have any other sampler worry about it, but I can see an argument for moving away from having the current `PrioritySampler` try to do it all.

I'm open to suggestions for different approaches to simplify this approach.

The current implementation in this PR works, but trying to maintain compatibility with all existing samplers, alongside the new rule sampler, has created an arms-race for the `PrioritySampler` class. I believe this implementation will be much simpler if we move away from having an agent one day.

Comment (delner) [reply lib/ddtrace/sampler.rb]: In the interest of time, we might want to keep what we have here if its working, and refactor in the future. I would just make sure you have sufficient tests to verify all these edge cases and conditions introduced by this complexity such that when we do refactor later, we won't introduce any bugs to the sampling logic. If we can do that, then I think its okay to let this be.

File: lib/ddtrace/sampler.rb (hunk 16 above)
Comment (delner) [reply lib/ddtrace/sampler.rb]: Gotcha, makes sense.

File: lib/ddtrace/sampling/matcher.rb [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+      attr_reader :name, :service
+
+      # @param name [String,Regexp,Proc] Matcher for case equality (===) with the span name, defaults to always match
+      # @param service [String,Regexp,Proc] Matcher for case equality (===) with the service name, defaults to always match
+      def initialize(name: MATCH_ALL, service: MATCH_ALL)
+        @name = name
+        @service = service
+      end
+
+      # rubocop:disable Style/CaseEquality
```
Comment (delner) [inline lib/ddtrace/sampling/matcher.rb]: This is a silly Rubocop rule... fine that you disabled it here. Might want to disable it in the config later too.

File: lib/ddtrace/sampling/rule.rb [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+  module Sampling
+    # Sampling rule that dictates if a span matches
+    # a specific criteria and what sampling strategy to
+    # apply in case of a positive match.
+    class Rule
+      extend Forwardable
+
+      attr_reader :matcher, :sampler
+
+      # @param [Matcher] matcher A matcher to verify span conformity against
```
Comment (delner) [inline lib/ddtrace/sampling/rule.rb]: I'm loving these comments you added to these functions! :100:

File: lib/ddtrace/sampling/token_bucket.rb (hunk 7 above)
Comment (delner) [reply lib/ddtrace/sampling/token_bucket.rb]: Alright, its probably fine using subclasses here, too. Just wanted to consider our options, but I'm cool with this.

**Comment by marcotc** [comment]
@delner I added integration tests recently to this PR, they are under [spec/ddtrace/integration_spec.rb](https://github.com/DataDog/dd-trace-rb/pull/854/files#diff-847f5f50f50bbcc3092347cf2381ec83).

**Comment by marcotc** [comment]
@delner added test coverage for the `sampler.rb` changes. Everything introduced in this PR should be covered now.

I added a few extra tests to cover `Sampler#sampling_rate` that were missing for existing functionality.

File: lib/ddtrace/sampling/matcher.rb (hunk 18 above)
Comment (marcotc) [reply lib/ddtrace/sampling/matcher.rb]: Done! Added to `.rubocop.yml`.

File: lib/ddtrace/sampling/rule_sampler.rb [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+                     default_sampler: nil)
+
+        @rules = rules
+
+        @rate_limiter = if rate_limiter
+                          rate_limiter
+                        elsif rate_limit
+                          Datadog::Sampling::TokenBucket.new(rate_limit)
+                        else
+                          Datadog::Sampling::TokenBucket.new(100)
```
Comment (brettlangdon) [inline lib/ddtrace/sampling/rule_sampler.rb]: The default here should be no rate limiter.

File: lib/ddtrace/sampler.rb [hunk 21]
```diff
@@ -229,8 +237,33 @@ def priority_assigned_upstream?(span)
...
+        # NOTE: We'll want to leave `span.sampled = true` here; all spans for priority sampling must
+        #       be sent to the agent. Otherwise metrics for traces will not be accurate, since the
+        #       agent will have an incomplete dataset.
+        #
+        #       We also ensure that the agent knows we that our `post_sampler` is not performing true sampling,
+        #       to avoid erroneous metric upscaling.
+        span.sampled = true
+        if pre_sample_rate_metric
+          # Restore true sampling metric, as only the @pre_sampler can reject traces
+          span.set_metric(SAMPLE_RATE_METRIC_KEY, pre_sample_rate_metric)
```
Comment (brettlangdon) [inline lib/ddtrace/sampler.rb]: This shouldn't be unset by any sampler right?

Comment (brettlangdon) [reply lib/ddtrace/sampler.rb]: It also shouldn't be set by any other sampler either

Comment (brettlangdon) [reply lib/ddtrace/sampler.rb]: This is only needed for samplers which cause traces to not be send to the agent `span.sampled = false`

Comment (marcotc) [reply lib/ddtrace/sampler.rb]: The default "post_sampler", `RateByServiceSampler`: https://github.com/DataDog/dd-trace-rb/blob/94b47a66ce57ff1e692920c320fc8040b3e63366/lib/ddtrace/sampler.rb#L192
ultimately delegate the sampling decision to an instance of `RateSampler`. That sampler sets this metric whenever it successfully samples: https://github.com/DataDog/dd-trace-rb/blob/94b47a66ce57ff1e692920c320fc8040b3e63366/lib/ddtrace/sampler.rb#L71-L73

In case case we restore it to the true sampling metric value, store under `pre_sample_rate_metric`.
No sampler today is unsetting this metric, but code is put in place for completeness.

#### Backstory
Ultimately, the current contract interface we have for samplers (`sample?` for no side-effects, and `sample!` for true sampling decision with side-effects) does not correctly allow the samplers to be used as both: 
1. Top-level sampler: which ultimately makes the sampling decision, possibly setting metrics and tags.
2. Chained sampler: provide a decision, but without side-effects. One example is the  "post_sampler" in `PrioritySampler` above.

We have cases where we want "some" side-effects, like in the `RuleSampler` when used in a sampler chain: it needs to set rule-specific metrics (e.g. rule sampling ratio metric), but not actually set the sampling decision. It's a "middle-ground" between the `sample!` and `sample?` of today.
I chose to go with an implementation with reduce changes to the existing samplers because:
* I did not have a clear vision on how to have a clean contract to represent all these requirements I mentioned above in a concise package.
* I did not want to perform a major refactoring on the existing samplers, introduction risk to users of the existing sampler.
